### PR TITLE
Add tests for stop server

### LINKED ISSUE #587: Stop server
Add new test for the server stop feature

### PR DESCRIPTION
**Pull request by whisperity**
Adds tests for #496. Closes #587.

### CODE REVIEW

File: tests/functional/instance_manager/test_instances.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+        self.assertTrue(any([True for p
+                             in instance_ports
+                             if p == codechecker_2['viewer_port']]),
+                        "Missing port for server started later!")
+
+        self.assertEqual(len(set(instance_ports)), 2,
+                         "The servers started with the same port.")
+
+    # NOTE: Do NOT rename this method. It MUST come lexicographically
```
Comment (gyorb) [inline tests/functional/instance_manager/test_instances.py]: I would recommend to move this note into the method documentation. Not before it.

File: tests/functional/instance_manager/test_instances.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        codechecker_1 = test_cfg['codechecker_1']
+        codechecker_2 = test_cfg['codechecker_2']
+        start_server(codechecker_2, test_cfg, 2)
+
+        self.assertEqual(len(instance_manager.list()), 2,
+                         "The started server was not appended to the "
+                         "instance list.")
+
+        # Workspaces must match, servers were started in the same workspace.
+        instance_wses = [i['workspace'] for i in instance_manager.list()]
```
Comment (gyorb) [inline tests/functional/instance_manager/test_instances.py]: This is a bit weird abbreviation for me `wses` for workspaces but I think I can live with this.

File: tests/functional/instance_manager/test_instances.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        self.assertEqual(instance_manager.list()[0]['port'],
+                         codechecker_1['viewer_port'],
+                         "The written port in the instance list is not the "
+                         "one the server was started with!")
+
+        self.assertEqual(instance_manager.list()[0]['workspace'],
+                         self._test_workspace,
+                         "Server was brought up with invalid workspace.")
+
+    # NOTE: Yet again keep the lexicographical flow...
```
Comment (gyorb) [inline tests/functional/instance_manager/test_instances.py]: The same here, please move the note into the method doc.

File: tests/functional/instance_manager/__init__.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    # TODO If environment variable is set keep the workspace
+    # and print out the path.
+    global TEST_WORKSPACE
+
+    print("Removing: " + TEST_WORKSPACE)
+    shutil.rmtree(TEST_WORKSPACE)
+
+
+def set_event(stop_event_idx):
```
Comment (gyorb) [inline tests/functional/instance_manager/__init__.py]: Maybe a `set_stop_event` name will be a bit more descriptive about what it does.

File: tests/functional/instance_manager/__init__.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import uuid
+from subprocess import CalledProcessError
+
+from libtest import get_free_port
+from libtest import project
+from libtest import codechecker
+from libtest import env
+
+# Stopping events for CodeChecker servers.
+__SERVER_EVENTS = [multiprocessing.Event(),
```
Comment (gyorb) [inline tests/functional/instance_manager/__init__.py]: Maybe we could use separate variable names like `server1_stop` for each stop event. 
Which make the stopping and starting more explicit and harder to mix the events.
`start_server(codechecker_1, test_cfg, server1_stop)` and to stop `set_event(server1_stop)`

File: tests/functional/instance_manager/__init__.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        'run_names': [],
+        'checkers': []
+    }
+
+    codechecker_cfg.update(host_port_cfg)
+
+    test_config['codechecker_1'] = codechecker_cfg
+
+    # We need a second server
+    codechecker_cfg = {
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/__init__.py]: Maybe we could somehow avoid repeating this literal?

File: tests/functional/instance_manager/test_instances.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+
+        test_class = self.__class__.__name__
+        print('Running ' + test_class + ' tests in ' + self._test_workspace)
+
+    def testInitialFile(self):
+        """Tests that initially az instance file is created."""
+
+        self.assertEqual(instance_manager.list(), [],
+                         instance_manager.list())
+        # "The initial instance file was not empty initially.")
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Why is there a closing paren in the comment?

File: tests/functional/instance_manager/test_instances.py [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+
+    def testServerStart(self):
+        """Started server writes itself to instance list"""
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Missing period.

File: tests/functional/instance_manager/test_instances.py [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+                         codechecker_1['viewer_port'],
+                         "The written port in the instance list is not the "
+                         "one the server was started with!")
+
+        self.assertEqual(instance_manager.list()[0]['workspace'],
+                         self._test_workspace,
+                         "Server was brought up with invalid workspace.")
+
+    def testServerStartSecondary(self):
+        """Another started server appends itself to instance list"""
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Missing period in the comment.

File: tests/functional/instance_manager/test_instances.py [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+        instance_workspaces = [i['workspace'] for i in instance_manager.list()]
+        self.assertEqual(instance_workspaces[0], self._test_workspace,
+                         "Server was brought up with invalid workspace.")
+        self.assertEqual(instance_workspaces[0], instance_workspaces[1],
+                         "Server was brought up with invalid workspace.")
+
+        # Exactly one server should own each port generated
+        instance_ports = [i['port'] for i in instance_manager.list()]
+
+        self.assertTrue(any([True for p
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Isn't there a way to express this in a more concise way? Like` codechecker_1['viewer_port'] in instance_ports` and also the next check.

File: tests/functional/instance_manager/test_instances.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+                        "Missing port for server started later!")
+
+        self.assertEqual(len(set(instance_ports)), 2,
+                         "The servers started with the same port.")
+
+    def testShutdownRecordKeeping(self):
+        """Test that one server's shutdown keeps the other records."""
+
+        # NOTE: Do NOT rename this method. It MUST come lexicographically
+        # AFTER testServerStartSecondary
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Why?

File: tests/functional/instance_manager/test_instances.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+
+        self.assertEqual(len(instance_manager.list()), 2,
+                         "The server disappeared from the instance list "
+                         "before shutdown.")
+
+        # Kill the second started server
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Missing periods.

File: tests/functional/instance_manager/test_instances.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+        self.assertEqual(0, run_cmd([env.codechecker_cmd(),
+                                     'server', '--stop',
+                                     '--view-port',
+                                     str(codechecker_2['viewer_port']),
+                                     '--workspace',
+                                     self._test_workspace]),
+                         "The stop command didn't return exit code 0.")
+        time.sleep(5)
+
+        # Check if the remaining server is still there
```
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Missing period.

File: tests/functional/instance_manager/test_instances.py (hunk 11 above)
Comment (whisperity) [inline tests/functional/instance_manager/test_instances.py]: If you look at the code, this method shuts down a server. To shut down a server, and to test that the shutdown actually does what it is expected to do, you need a server to be running before you attempt to shut it down.

And _nose_ executes tests in a lexicographical order.

File: tests/functional/instance_manager/__init__.py (hunk 6 above)
Comment (whisperity) [inline tests/functional/instance_manager/__init__.py]: We need two unaliased copies of this dict in the `test_config` as they need to configure two separate servers.

Comment (Xazax-hun) [inline tests/functional/instance_manager/__init__.py]: You could do a deep copy to make it unaliased.

File: tests/functional/instance_manager/test_instances.py (hunk 11 above)
Comment (Xazax-hun) [inline tests/functional/instance_manager/test_instances.py]: Maybe this last piece of information about nose should be in the comment.

File: tests/functional/instance_manager/__init__.py (hunk 6 above)
Comment (whisperity) [inline tests/functional/instance_manager/__init__.py]: It's either this, or

```python
import copy
# ...
copy.deepcopy()
```

But we don't need to deep-copy the remaining part of the record, `TEST_WORKSPACE` can alias, for example. But the later appended `host_port_cfg` can not.

Comment (Xazax-hun) [inline tests/functional/instance_manager/__init__.py]: I'd prefer the deepcopy, but I can accept your view.

**Pull request merged**